core/src/main/java/io/grpc/Status.java

```diff
@@ -571,15 +569,93 @@ public String toString() {
         .toString();
   }
 
-  private static class StatusCodeMarshaller implements Metadata.AsciiMarshaller<Status> {
+  private static final class StatusCodeMarshaller implements TrustedAsciiMarshaller<Status> {
     @Override
-    public String toAsciiString(Status status) {
+    public byte[] toAsciiString(Status status) {
       return status.getCode().valueAscii();
     }
 
     @Override
-    public Status parseAsciiString(String serialized) {
-      return fromCodeValue(Integer.valueOf(serialized));
+    public Status parseAsciiString(byte[] serialized) {
+      return fromCodeValue(serialized);
+    }
+  }
+
+  private static final class StatusMessageMarshaller implements TrustedAsciiMarshaller<String> {
+
+    private static final byte[] HEX =
+        {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+    @Override
+    public byte[] toAsciiString(String value) {
+      byte[] valueBytes = value.getBytes(UTF_8);
```

Comment:
Arbitrary unicode is allowed as input.  The conversion process takes arbitrary user input string, converts it into UTF-8 bytes, and then escapes all non ascii printable bytes.  This allows it to be ASCII on the wire, with the expectation that most status messages are from the ASCII character space.